src/jit/codegenarm64.cpp

```diff
@@ -2665,62 +2665,191 @@ void CodeGen::genJumpTable(GenTree* treeNode)
     regNumber addrReg   = addr->gtRegNum;
-    instruction ins;
 
-    // all of these nodes implicitly do an indirection on op1
-    // so create a temporary node to feed into the pattern matching
-    GenTreeIndir i = indirForm(data->TypeGet(), addr);
-    genConsumeReg(addr);
+    regNumber exResultReg  = treeNode->ExtractTempReg(RBM_ALLINT);
+    regNumber storeDataReg = (treeNode->OperGet() == GT_XCHG) ? dataReg : treeNode->ExtractTempReg(RBM_ALLINT);
+    regNumber loadReg      = (targetReg != REG_NA) ? targetReg : storeDataReg;
 
     // The register allocator should have extended the lifetime of the address
     // so that it is not used as the target.
-    noway_assert(addrReg != targetReg);
+    assert(addrReg != targetReg);
 
-    // If data is a lclVar that's not a last use, we'd better have allocated a register
-    // for the result (except in the case of GT_LOCKADD which does not produce a register result).
-    assert(targetReg != REG_NA || treeNode->OperGet() == GT_LOCKADD || !genIsRegCandidateLocal(data) || (data->gtFlags & GTF_VAR_DEATH) != 0);
+    assert(addrReg != loadReg);
+    assert(dataReg != loadReg);
 
-    genConsumeIfReg(data);
-    if (targetReg != REG_NA && dataReg != REG_NA && dataReg != targetReg)
-    {
-        inst_RV_RV(ins_Copy(data->TypeGet()), targetReg, dataReg);
-        data->gtRegNum = targetReg;
+    assert(addrReg != storeDataReg);
+    assert((treeNode->OperGet() == GT_XCHG) || (addrReg != dataReg));
+
+    assert(addr->isUsedFromReg());
+    assert(exResultReg != REG_NA);
+    assert(exResultReg != targetReg);
+    assert((targetReg != REG_NA) || (treeNode->OperGet() != GT_XCHG));
+
+    // Store exclusive unpredictable cases must be avoided
+    assert(exResultReg != storeDataReg);
+    assert(exResultReg != addrReg);
+
+    genConsumeAddress(addr);
+    genConsumeRegs(data);
+
+    // NOTE: `genConsumeAddress` marks the consumed register as not a GC pointer, as it assumes that the input registers
+    // die at the first instruction generated by the node. This is not the case for these atomics as the  input
+    // registers are multiply-used. As such, we need to mark the addr register as containing a GC pointer until
+    // we are finished generating the code for this node.
+
+    gcInfo.gcMarkRegPtrVal(addrReg, addr->TypeGet());
+
+    // TODO-ARM64-CQ Use ARMv8.1 atomics if available
+    // https://github.com/dotnet/coreclr/issues/11881
+
+    // Emit code like this:
+    //   retry:
+    //     ldxr loadReg, [addrReg]
+    //     add storeDataReg, loadReg, dataReg         # Only for GT_XADD & GT_LOCKADD
+    //                                                # GT_XCHG storeDataReg === dataReg
+    //     stxr exResult, storeDataReg, [addrReg]
+    //     cbnz exResult, retry
+
+    BasicBlock* labelRetry = genCreateTempLabel();
+    genDefineTempLabel(labelRetry);
+
+    getEmitter()->emitIns_R_R(INS_ldaxr, emitTypeSize(treeNode), loadReg, addrReg);
```

Comment:
Acquire half barrier

Reply:
Are going to add this as a comment here?

Reply:
Opened an issue an added comments